### PR TITLE
Serialize all foreign `SourceFile`s into proc-macro crate metadata

### PR DESCRIPTION
**Pull request by Aaron1011**
Normally, we encode a `Span` that references a foreign `SourceFile` by
encoding information about the foreign crate. When we decode this
`Span`, we lookup the foreign crate in order to decode the `SourceFile`.

However, this approach does not work for proc-macro crates. When we load
a proc-macro crate, we do not deserialzie any of its dependencies (since
a proc-macro crate can only export proc-macros). This means that we
cannot serialize a reference to an upstream crate, since the associated
metadata will not be available when we try to deserialize it.

This commit modifies foreign span handling so that we treat all foreign
`SourceFile`s as local `SourceFile`s when serializing a proc-macro.
All `SourceFile`s will be stored into the metadata of a proc-macro
crate, allowing us to cotinue to deserialize a proc-macro crate without
needing to load any of its dependencies.

Since the number of foreign `SourceFile`s that we load during a
compilation session may be very large, we only serialize a `SourceFile`
if we have also serialized a `Span` which requires it.

### CODE REVIEW

**Comment by Aaron1011** [comment]
I think this code doesn't need to be changed, but I'm not 100% certain:

https://github.com/rust-lang/rust/blob/d8ed1b03c202d77248eb0d335062f46026fc29c4/src/librustc_middle/hir/map/collector.rs#L171-L182

When we encode a proc-macro crate, we'll end up serializing some foreign `SourceFile`s, which we won't hash. However, we will still include our dependency hashes in the crate hash, which will cause those `SourceFiles` to influence the final hash.

**Comment by petrochenkov** [comment]
>I think this code doesn't need to be changed, but I'm not 100% certain

I'm also not sure, but the logic seems reasonable.

File: src/librustc_metadata/rmeta/encoder.rs [hunk 1]
```diff
@@ -154,18 +168,24 @@ impl<'tcx> SpecializedEncoder<Span> for EncodeContext<'tcx> {
...
-        if !self.source_file_cache.contains(span.hi) {
+        if !self.source_file_cache.0.contains(span.hi) {
             // Unfortunately, macro expansion still sometimes generates Spans
             // that malformed in this way.
             return TAG_INVALID_SPAN.encode(self);
         }
 
+        let source_files = self.required_source_files.as_mut().expect("Already encoded SourceMap!");
+        // Record the fact that we need to encode the data for this `SourceFile`
+        source_files.ensure(self.source_file_cache.1 + 1);
```
Comment (petrochenkov) [inline src/librustc_metadata/rmeta/encoder.rs:186]: Nit: looks like `ensure` is called by `insert` automatically.

**Comment by petrochenkov** [comment]
Something like this is definitely preferable to https://github.com/rust-lang/rust/pull/69976 as and end state.

r=me with the nit addressed

**Comment by Aaron1011** [comment]
I've removed the call to `ensure`.

@bors r=petrochenkov

**Comment by Manishearth** [comment]
@bors r-

failed the rollup, likely

https://github.com/rust-lang/rust/pull/73824#issuecomment-650691210

https://dev.azure.com/rust-lang/rust/_build/results?buildId=33134&view=logs&j=1c64c3ff-4c37-56fe-a133-b407354bbfbf&t=4d85e1cd-e19c-562d-4a04-e8daca7cfc12&l=17311

<details>

```
failures:

---- [ui] ui/proc-macro/meta-macro-hygiene.rs stdout ----
diff of stdout:
#1: parent: #0, outer_mark: (ExpnId(1), Opaque)
#2: parent: #0, outer_mark: (ExpnId(1), Transparent)
#3: parent: #0, outer_mark: (ExpnId(2), Opaque)
#4: parent: #0, outer_mark: (ExpnId(2), Transparent)
#5: parent: #0, outer_mark: (ExpnId(2), SemiTransparent)
*/

------------------------------------------
stderr:
------------------------------------------

------------------------------------------



failures:
    [ui] ui/proc-macro/meta-macro-hygiene.rs

test result: FAILED. 10350 passed; 1 failed; 62 ignored; 0 measured; 0 filtered out
```

</details>

**Comment by Aaron1011** [comment]
@bors r+ rollup=never

It looks like one of the other PRs in the rollup affected a symbol ID in src/test/ui/proc-macro/meta-macro-hygiene.rs

**Comment by Aaron1011** [comment]
It looks like the `Symbol` ids are different across platforms?

**Comment by Aaron1011** [comment]
Forgot to bless a stderr file

@bors r=petrochenkov

**Comment by Manishearth** [comment]
@bors r- yield retry

sorry, _just_ made a rollup

**Comment by mati865** [comment]
Nice performance improvement :tada:: https://perf.rust-lang.org/compare.html?start=16957bd4d3a5377263f76ed74c572aad8e4b7e59&end=d462551a8600e57d8b6f87e71ea56868bc5da6cf&stat=instructions:u

**Pull request merged**